connectors/utils.py

```diff
@@ -350,6 +350,20 @@ def put_nowait(self, item):
         super().put_nowait((item_size, item))
 
 
+class NonBlockingBoundedSemaphore(asyncio.BoundedSemaphore):
```

Comment:
I'm really against inheritance here - why not make it a method in the caller (ConcurrentTasks)?

Additionally, in this PR method is not used at all, why?

Reply:
>I'm really against inheritance here - why not make it a method in the caller (ConcurrentTasks)?

Can you elaborate more? 

>Additionally, in this PR method is not used at all, why?

It will be used by PR https://github.com/elastic/connectors/pull/1479

The implementation is based on the discussion [here](https://github.com/elastic/connectors/pull/1479#discussion_r1383112200)

Reply:
> Can you elaborate more?

I thought what's bothering me and left a comment about private variable usage.

Overall it feels overkill to have inheritance here - we just add a method to the class, when in reality it's more of a helper method. I like to use inheritance only when it's _needed_ - things like polymorphism, double/single dispatch patterns, marker interfaces and regular interfaces, etc.

I've seen ES client do helper methods like this: https://elasticsearch-py.readthedocs.io/en/v7.12.0/helpers.html

So in our case it would be:

```python
# connectors/utils.py
def try_acquire(semaphore):
  # do things

# actual code
result = try_acquire(semaphore)
```

Reply:
>It will be used by PR https://github.com/elastic/connectors/pull/1479
Got it! Thanks for the explanation, but this process still confuses me a bit - it's much easier to review code when code that's added is used in the PR.

Reply:
> I've seen ES client do helper methods like this: https://elasticsearch-py.readthedocs.io/en/v7.12.0/helpers.html
> 
> So in our case it would be:

It's not guaranteed to work. See the experiment [here](https://github.com/elastic/connectors/pull/1479#discussion_r1387783096). as long as there's an `await`, context switching may happen.

>it's much easier to review code when code that's added is used in the PR.

But sometimes I also want to break a large PR into smaller ones for easy review. WDYT if we add new functions with clear explanation of the context? Even though it's not used immediately, the PR will make sure the function is working as expected with tests.